File: path_sum_iv/Path_Sum_IV.cpp
class Path_Sum_IV {
public:
    int pathSum(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }

        unordered_map<int, unordered_map<int, int>> nodes;
        for (int num : nums) {
            int lvl = num / 100;
            int idx = (num / 10) % 10 - 1;
            int val = num % 10;
            nodes[lvl][idx] = val;
        }

        int sum = 0;
        dfs(nodes, 1, 0, 0, sum);
        return sum;
    }

private:
    void dfs(unordered_map<int, unordered_map<int, int>>& nodes, int lvl, int idx, int curr_sum, int& sum) {
        curr_sum += nodes[lvl][idx];
        if (nodes.find(lvl + 1) == nodes.end() || (nodes[lvl + 1].find(2 * idx) == nodes[lvl + 1].end() && nodes[lvl + 1].find(2 * idx + 1) == nodes[lvl + 1].end())) {
            sum += curr_sum;
            return;
        }

        if (nodes[lvl + 1].find(2 * idx) != nodes[lvl + 1].end()) {
            dfs(nodes, lvl + 1, 2 * idx, curr_sum, sum);
        }
        if (nodes[lvl + 1].find(2 * idx + 1) != nodes[lvl + 1].end()) {
            dfs(nodes, lvl + 1, 2 * idx + 1, curr_sum, sum);
        }
    }
};

File: path_sum_iv/PathSumIV.java
class PathSumIV {
    public int pathSum(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        Map<Integer, Map<Integer, Integer>> nodes = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int[] details = getDetails(nums[i]);
            if (!nodes.containsKey(details[0])) {
                nodes.put(details[0], new HashMap<>());
            }
            nodes.get(details[0]).put(details[1] - 1, details[2]);
        }

        int[] sum = new int[]{0};
        dfs(nodes, 1, 0, 0, sum);
        return sum[0];
    }

    private void dfs(Map<Integer, Map<Integer, Integer>> nodes, int x, int y, int currSum, int[] sum) {
        currSum += nodes.get(x).get(y);
        if (!nodes.containsKey(x + 1) || (!nodes.get(x + 1).containsKey(2 * y) && !nodes.get(x + 1).containsKey(2 * y + 1))) {
            sum[0] += currSum;
            return;
        }
        
        if (nodes.get(x + 1).containsKey(2 * y)) {
            dfs(nodes, x + 1, 2 * y, currSum, sum);
        }
        if (nodes.get(x + 1).containsKey(2 * y + 1)) {
            dfs(nodes, x + 1, 2 * y + 1, currSum, sum);
        }
    }

    private int[] getDetails(int num) {
        int[] details = new int[3];
        for (int i = 2; i >= 0; i--) {
            details[i] = num % 10;
            num /= 10;
        }

        return details;
    }
}